File: src/processes/tests/halo/gov.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import chalk from "chalk";
import * as chai from "chai";
import chaiAsPromised from "chai-as-promised";
import { SigningCosmWasmClient } from "@cosmjs/cosmwasm-stargate";
import { sendTransaction, toEncodedBinary } from "../../../utils/helpers";

chai.use(chaiAsPromised);
const { expect } = chai;

export enum VoteOption {
  YES,
  NO,
}

//----------------------------------------------------------------------------------------
// TEST: Update gov config
//
// SCENARIO:
// Pleb cannot update contract config, only owner can update config
//
//----------------------------------------------------------------------------------------
export async function testGovUpdateConfig(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  owner: string | undefined,
  quorum: number | undefined,
  threshold: number | undefined,
  voting_period: number | undefined,
  timelock_period: number | undefined,
  proposal_deposit: string | undefined,
  snapshot_period: number | undefined,
  unbonding_period: number | undefined,
  gov_hodler: string | undefined
): Promise<void> {
  process.stdout.write("Test - Only owner can update gov config");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      update_config: {
        owner,
        quorum,
        threshold,
        voting_period,
        timelock_period,
        proposal_deposit,
        snapshot_period,
        unbonding_period,
        gov_hodler,
      },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Update the configs of the Gov Hodler Contract
//----------------------------------------------------------------------------------------
export async function testGovHodlerUpdateConfig(
  juno: SigningCosmWasmClient,
  apTeam: string,
  gov_hodler: string,
  owner: string | undefined,
  govContract: string | undefined
): Promise<void> {
  process.stdout.write("Test - Only owner can update gov hodler config");

  await expect(
    sendTransaction(juno, apTeam, gov_hodler, {
      update_config: {
        owner,
        gov_contract: govContract,
      },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Transfer Stake from Old to New Gov Contract
//----------------------------------------------------------------------------------------
export async function testTransferStake(
  juno: SigningCosmWasmClient,
  apTeam: string,
  oldGov: string,
  newGov: string,
  staker_info: string[][]
): Promise<void> {
  process.stdout.write("Test - Execute transfer of staker balances to a new contract");
  let msgs: Msg[] = [];
  staker_info.forEach((info) => {
    msgs.push(
      new MsgExecuteContract(apTeam, oldGov, {
        transfer_stake: {
          new_gov_contract: newGov,
          address: info[0],
          amount: info[1],
        },
      })
    );
  });

  await expect(sendTransaction(juno, apTeam, msgs));
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: End a poll
//
// SCENARIO:
// End a poll
//
//----------------------------------------------------------------------------------------
export async function testGovEndPoll(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  poll_id: number
): Promise<void> {
  process.stdout.write("Test - Execute a msgs of poll");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      end_poll: { poll_id },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Execute a poll
//
// SCENARIO:
// User can execute a poll that has passed vote
// User cannot execute a poll that has not passed or expired vote
//
//----------------------------------------------------------------------------------------
export async function testGovExecutePoll(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  poll_id: number
): Promise<void> {
  process.stdout.write("Test - Execute a poll");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      execute_poll: { poll_id },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Snapshot a poll
//
// SCENARIO:
// SnapshotPoll is used to take a snapshot of the staked amount for quorum calculation
//
//----------------------------------------------------------------------------------------
export async function testGovSnapshotPoll(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  poll_id: number
): Promise<void> {
  process.stdout.write("Test - Execute a poll");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      snapshot_poll: { poll_id },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Register contracts
//
// SCENARIO:
// Register contracts
//
//----------------------------------------------------------------------------------------
export async function testGovRegisterContracts(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  halo_token: string
): Promise<void> {
  process.stdout.write("Test - Gov register staking token contract");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      register_contracts: { halo_token },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Stake HALO voting tokens
//
// SCENARIO:
// Stake some amount of HALO tokens for voting on Polls
//
//----------------------------------------------------------------------------------------
export async function testGovStakeVotingTokens(
  juno: SigningCosmWasmClient,
  apTeam: string,
  haloToken: string,
  govContract: string,
  amount: string
): Promise<void> {
  process.stdout.write("Test - Stake voting tokens");

  await expect(
    sendTransaction(juno, apTeam, haloToken, {
      send: {
        amount: amount,
        contract: govContract,
        msg: toEncodedBinary({ stake_voting_tokens: {} }),
      },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Withdraw voting tokens
//
// SCENARIO:
// Withdraw amount if not staked. By default all funds will be withdrawn.
// Withdrawn HALO is subject to an unbonding period before it can be accessed.
//
//----------------------------------------------------------------------------------------
export async function testGovWithdrawVotingTokens(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  amount: string | undefined
): Promise<void> {
  process.stdout.write("Test - Withdraw voting tokens");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      withdraw_voting_tokens: { amount },
    })
  );
  console.log(chalk.green(" Passed!"));
}

export async function testGovClaimVotingTokens(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string
): Promise<void> {
  process.stdout.write("Test - Claim all eligable withdrawn voting tokens");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      claim_voting_tokens: {},
    })
  );
  console.log(chalk.green(" Passed!"));
}

export async function testGovResetClaims(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  addresses: string[]
): Promise<void> {
  process.stdout.write("Test - Reset claims for all addresses passed");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      reset_claims: { claim_addrs: addresses },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Cast vote
//
// SCENARIO:
// AP Team/User can vote on the open poll
//
//----------------------------------------------------------------------------------------
export async function testGovCastVote(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  poll_id: number,
  vote: VoteOption,
  amount: string
): Promise<void> {
  process.stdout.write("Test - Cast vote");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      cast_vote: { poll_id, vote, amount },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// TEST: Execute a poll
//
// SCENARIO:
// Submit a poll for changing registrar settings
//
//----------------------------------------------------------------------------------------
export async function testGovExecutePollForRegistrarSettings(
  juno: SigningCosmWasmClient,
  apTeam: string,
  govContract: string,
  halo_token: string,
  funding_goal: string | undefined,
  fund_rotation: number | undefined,
  split_to_liquid: string | undefined,
  treasury_tax_rate: string | undefined
): Promise<void> {
  process.stdout.write("Test - Execute a poll");

  await expect(
    sendTransaction(juno, apTeam, govContract, {
      receive: {
        sender: apTeam,
        amount: "123",
        msg: toEncodedBinary({
          title: "title",
          description: "description",
          link: undefined,
          proposal_type: "registrar",
          options: [
            {
              order: 1,
              funding_goal,
              fund_rotation,
              split_to_liquid,
              treasury_tax_rate,
              msg: toEncodedBinary({ amount: "123" }),
            },
          ],
        }),
      },
    })
  );
  console.log(chalk.green(" Passed!"));
}

//----------------------------------------------------------------------------------------
// Querying tests
//----------------------------------------------------------------------------------------
export async function testQueryGovConfig(
  juno: SigningCosmWasmClient,
  govContract: string
): Promise<void> {
  process.stdout.write("Test - Query Gov Config");
  const result: any = await juno.queryContractSmart(govContract, {
    config: {},
  });

  console.log(result);
  console.log(chalk.green(" Passed!"));
}

export async function testQueryGovState(
  juno: SigningCosmWasmClient,
  govContract: string
): Promise<void> {
  process.stdout.write("Test - Query Gov State");
  const result: any = await juno.queryContractSmart(govContract, {
    state: {},
  });

  console.log(result);
  console.log(chalk.green(" Passed!"));
}

export async function testQueryGovStaker(
  juno: SigningCosmWasmClient,
  govContract: string,
  address: string
): Promise<void> {
  process.stdout.write("Test - Query for getting gov staker");
  const result: any = await juno.queryContractSmart(govContract, {
    staker: { address },
  });

  console.log(result);
  console.log(chalk.green(" Passed!"));
}

export async function testQueryGovClaims(
  juno: SigningCosmWasmClient,
  govContract: string,
  address: string
): Promise<void> {
  process.stdout.write("Test - Query Gov Claims for an addr");
  const result: any = await juno.queryContractSmart(govContract, {
    claims: { address },
  });

  // console.log(result);
  result.claims.forEach((r: any) => console.log(r, r.release_at));
  console.log(chalk.green(" Passed!"));
}

export async function testQueryGovPoll(
  juno: SigningCosmWasmClient,
  govContract: string,
  poll_id: number
): Promise<void> {
  process.stdout.write("Test - Query for getting poll by poll_id");
  const result: any = await juno.queryContractSmart(govContract, {
    poll: { poll_id },
  });

  console.log(result);
  console.log(chalk.green(" Passed!"));
}

export async function testQueryGovPolls(
  juno: SigningCosmWasmClient,
  govContract: string,
  filter: any | undefined,
  start_after: string | undefined,
  limit: number | undefined
): Promise<void> {
  process.stdout.write("Test - Query for getting poll by poll_id");
  const result: any = await juno.queryContractSmart(govContract, {
    polls: { filter, start_after, limit, undefined },
  });

  console.log(result);
  console.log(chalk.green(" Passed!"));
}

export async function testQueryGovVoters(
  juno: SigningCosmWasmClient,
  govContract: string,
  poll_id: number,
  start_after: string | undefined,
  limit: number | undefined
): Promise<void> {
  process.stdout.write("Test - Query gov voters");
  const result: any = await juno.queryContractSmart(govContract, {
    voters: { poll_id, start_after, limit, undefined },
  });

  console.log(result);
  console.log(chalk.green(" Passed!"));
}

File: cosmwasm/packages/halo_token/src/gov.rs
use crate::common::OrderBy;
#[allow(unused_imports)]
use crate::staking::StakerInfoResponse;
use cosmwasm_schema::{cw_serde, QueryResponses};
use cosmwasm_std::{Binary, Decimal, Uint128};
use cw20::Cw20ReceiveMsg;
#[allow(unused_imports)]
use cw_controllers::{Claim, ClaimsResponse};
use cw_utils::Duration;
use std::fmt;

/// We currently take no arguments for migrations
#[cw_serde]
pub struct MigrateMsg {}

#[cw_serde]
pub struct InstantiateMsg {
    pub quorum: u64,
    pub threshold: u64,
    pub voting_period: u64,
    pub timelock_period: u64,
    pub proposal_deposit: Uint128,
    pub snapshot_period: u64,
    pub registrar_contract: String,
    pub halo_token: String,    // halo token address
    pub unbonding_period: u64, // days of unbonding
    pub gov_hodler: String,    // contract to hold maturing claims
}

#[cw_serde]
pub enum ExecuteMsg {
    Receive(Cw20ReceiveMsg),
    ExecutePollMsgs {
        poll_id: u64,
    },
    RegisterContracts {
        halo_token: String,
    },
    UpdateConfig {
        owner: Option<String>,
        quorum: Option<u64>,
        threshold: Option<u64>,
        voting_period: Option<u64>,
        timelock_period: Option<u64>,
        proposal_deposit: Option<Uint128>,
        snapshot_period: Option<u64>,
        unbonding_period: Option<u64>,
        gov_hodler: Option<String>,
    },
    CastVote {
        poll_id: u64,
        vote: VoteOption,
        amount: Uint128,
    },
    WithdrawVotingTokens {
        amount: Option<Uint128>,
    },
    ClaimVotingTokens {},
    EndPoll {
        poll_id: u64,
    },
    ExecutePoll {
        poll_id: u64,
    },
}

#[cw_serde]
pub enum Cw20HookMsg {
    /// StakeVotingTokens a user can stake their mirror token to receive rewards
    /// or do vote on polls
    StakeVotingTokens {},
    /// CreatePoll need to receive deposit from a proposer
    CreatePoll {
        title: String,
        description: String,
        link: Option<String>,
        proposal_type: Option<String>,
        options: Option<Vec<PollExecuteMsg>>,
    },
}

#[cw_serde]
pub struct PollExecuteMsg {
    pub order: u64,
    pub msg: Binary,
    pub funding_goal: Option<Uint128>,
    pub fund_rotation: Option<u64>,
    pub split_to_liquid: Option<Decimal>,
    pub treasury_tax_rate: Option<Decimal>,
}

#[cw_serde]
#[derive(QueryResponses)]
pub enum QueryMsg {
    #[returns(ConfigResponse)]
    Config {},
    #[returns(StateResponse)]
    State {},
    /// Claims shows the number of tokens this address can access when they are done unbonding
    #[returns(ClaimsResponse)]
    Claims { address: String },
    #[returns(StakerInfoResponse)]
    Staker { address: String },
    #[returns(PollResponse)]
    Poll { poll_id: u64 },
    #[returns(PollsResponse)]
    Polls {
        filter: Option<PollStatus>,
        start_after: Option<u64>,
        limit: Option<u32>,
        order_by: Option<OrderBy>,
    },
    #[returns(VotersResponse)]
    Voters {
        poll_id: u64,
        start_after: Option<String>,
        limit: Option<u32>,
        order_by: Option<OrderBy>,
    },
}

#[cw_serde]
pub struct ConfigResponse {
    pub owner: String,
    pub halo_token: String,
    pub quorum: Decimal,
    pub threshold: Decimal,
    pub voting_period: u64,
    pub timelock_period: u64,
    pub proposal_deposit: Uint128,
    pub snapshot_period: u64,
    pub unbonding_period: Duration,
}

#[cw_serde]
pub struct StateResponse {
    pub poll_count: u64,
    pub total_share: Uint128,
    pub total_deposit: Uint128,
}

#[cw_serde]
pub struct PollResponse {
    pub id: u64,
    pub creator: String,
    pub status: PollStatus,
    pub end_height: u64,
    pub title: String,
    pub description: String,
    pub link: Option<String>,
    pub proposal_type: Option<String>,
    pub deposit_amount: Uint128,
    pub execute_data: Option<Vec<PollExecuteMsg>>,
    pub yes_votes: Uint128, // balance
    pub no_votes: Uint128,  // balance
    pub staked_amount: Option<Uint128>,
    pub total_balance_at_end_poll: Option<Uint128>,
}

#[cw_serde]
pub struct PollsResponse {
    pub polls: Vec<PollResponse>,
}

#[cw_serde]
pub struct PollCountResponse {
    pub poll_count: u64,
}

#[cw_serde]
pub struct StakerResponse {
    pub balance: Uint128,
    pub share: Uint128,
    pub locked_balance: Vec<(u64, VoterInfo)>,
    pub claims: Vec<Claim>,
}

#[cw_serde]
pub struct VotersResponseItem {
    pub voter: String,
    pub vote: VoteOption,
    pub balance: Uint128,
}

#[cw_serde]
pub struct VotersResponse {
    pub voters: Vec<VotersResponseItem>,
}

#[cw_serde]
pub struct VoterInfo {
    pub vote: VoteOption,
    pub balance: Uint128,
}

#[cw_serde]
pub enum PollStatus {
    InProgress,
    Passed,
    Rejected,
    Executed,
    Expired, // Depricated
    Failed,
}

impl fmt::Display for PollStatus {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[cw_serde]
pub enum VoteOption {
    Yes,
    No,
}

impl fmt::Display for VoteOption {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        if *self == VoteOption::Yes {
            write!(f, "yes")
        } else {
            write!(f, "no")
        }
    }
}
